File: src/lib/base68.ts
const BASE68_CHARS =
  '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_*+!';
const BASE_LENGTH = BASE68_CHARS.length;

export function incrementBase68String(str: String) {
  if (str === '') return BASE68_CHARS[0];

  let indices = str.split('').map((char) => BASE68_CHARS.indexOf(char));
  let carry = 1;

  for (let i = indices.length - 1; i >= 0; i--) {
    if (carry === 0) break;

    let newVal = indices[i] + carry;
    carry = newVal >= BASE_LENGTH ? 1 : 0;
    indices[i] = newVal % BASE_LENGTH;
  }

  if (carry > 0) {
    indices.unshift(0);
  }

  return indices.map((i) => BASE68_CHARS[i]).join('');
}

File: src/base68/base68.go
package base68

import (
	"bytes"
	"fmt"
)

var (
	base68Chars = []byte("0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_~$!")
)

func IncrementBase68String(indexString string) string {
	index := []byte(indexString)
	if len(index) == 0 {
		return string([]byte{base68Chars[0]})
	}

	indices := make([]int, len(index))
	for i, char := range index {
		indices[i] = bytes.IndexByte(base68Chars, char)
	}

	for i := len(indices) - 1; i >= 0; i-- {

		newVal := indices[i] + 1

		indices[i] = newVal % len(base68Chars)
		fmt.Printf("%v, %v\n", newVal, len(base68Chars))
		if newVal < len(base68Chars) {
			break
		}

		if i == 0 {
			indices = append([]int{0}, indices...)
		}

	}

	result := make([]byte, len(indices))
	for i, idx := range indices {
		result[i] = base68Chars[idx]
	}
	return string(result)
}
